internal/rbd/nodeserver.go

```diff
@@ -1190,26 +1197,54 @@ func (ns *NodeServer) xfsSupportsReflink() bool {
 func (ns *NodeServer) NodeGetVolumeStats(
 	ctx context.Context,
 	req *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error) {
-	var err error
-	targetPath := req.GetVolumePath()
-	if targetPath == "" {
-		err = fmt.Errorf("targetpath %v is empty", targetPath)
+	var (
+		imgInfo rbdImageMetadataStash
+		err     error
+	)
+	volumePath := req.GetStagingTargetPath()
+	if volumePath == "" {
+		if volumePath = req.GetVolumePath(); volumePath == "" {
+			return nil, status.Error(codes.InvalidArgument, "volume path must be provided")
+		}
+	} else {
+		volumePath = volumePath + "/" + req.GetVolumeId()
+		imgInfo, err = lookupRBDImageMetadataStash(volumePath)
+		if err != nil {
+			log.ErrorLog(ctx, "failed to find image metadata, at stagingPath: %s, err: %v", volumePath, err)
 
-		return nil, status.Error(codes.InvalidArgument, err.Error())
+			return nil, status.Error(codes.NotFound, err.Error())
+		}
 	}
 
-	stat, err := os.Stat(targetPath)
+	var volCondition *csi.VolumeCondition
+	stat, err := os.Stat(volumePath)
 	if err != nil {
-		return nil, status.Errorf(codes.InvalidArgument, "failed to get stat for targetpath %q: %v", targetPath, err)
+		volCondition.Abnormal = true
+		volCondition.Message = fmt.Sprintf("failed to get stat for volumepath %q: %v", volumePath, err)
+	}
+
+	// userspace mounters
+	if imgInfo.NbdAccess {
+		stdout, stderr, err := util.ExecCommand(ctx, "pstree", "--arguments", "|", "grep", "[r]bd-nbd")
```

Comment:
This is rather ugly. Is there no way to validate that rbd-nbd is running for this particular volume?

Reply:
I know, this is just a handy way! I will see if we can fetch the image spec details with `rbd-nbd list-mapped` command.